todolist/client/src/main/scala/handlers/PingPongClientHandler.scala

```diff
@@ -17,21 +17,21 @@
 package examples.todolist.client
 package handlers
 
-import cats.syntax.functor._
+import cats.syntax.flatMap._
 import cats.effect.{Resource, Sync}
 import examples.todolist.client.clients.PingPongClient
 import examples.todolist.protocol.Protocols.PingPongService
 import higherkindness.mu.rpc.protocol.Empty
-import org.log4s._
+import io.chrisdavenport.log4cats.Logger
+import io.chrisdavenport.log4cats.slf4j.Slf4jLogger
 
 class PingPongClientHandler[F[_]: Sync](client: Resource[F, PingPongService[F]])
     extends PingPongClient[F] {
 
-  val logger: Logger = getLogger
+  implicit def unsafeLogger[L[_]: Sync] = Slf4jLogger.getLogger[F]
```

Comment:
The problem with this `implicit def unsafeLogger` is that it will result in a call to the underlying slf4j `getLogger` method every single time you call `Logger[F]`. That won't actually result in a new logger instance each time, because they are cached, but it's still a bit wasteful.

I think we have 2 options.

Either save the logger as a `val`:

```scala
class PingPongClientHandler[F[_]: Sync](client: Resource[F, PingPongService[F]])
    extends PingPongClient[F] {

  val logger: Logger[F] = Slf4jLogger.getLogger[F]

  override def ping(): F[Unit] =
    client
      .use(_.ping(Empty))	
      .flatMap(p => logger.info(s"Pong received with timestamp: ${p.time}"))
}
```

or create a logger once at the edge of the app (in `implicits.scala`?) and pass it in implicitly:

```scala
class PingPongClientHandler[F[_]: Sync: Logger](client: Resource[F, PingPongService[F]])
    extends PingPongClient[F] {

  val logger: Logger[F] = Slf4jLogger.getLogger[F]

  override def ping(): F[Unit] =
    client
      .use(_.ping(Empty))	
      .flatMap(p => Logger[F].info(s"Pong received with timestamp: ${p.time}"))
}
```

I would go for the first option as it's simpler and it will produce a logger that's properly named based on the class name. (`Slf4jLogger.getLogger` is actually a macro that looks at the enclosing class's name.)

Reply:
this is great feedback; thanks!  I'll make this change.